Add fine-grained `__has_feature()` cutout

j39m opened the pull request:
This is a follow-up to pull #148323. It mints
`-fsanitize-ignore-for-ubsan-feature=...`, accepting a list of (UBSan) sanitizers that should not cause
`__has_feature(undefined_behavior_sanitizer)` to evaluate true.

Commit:
Add fine-grained `__has_feature()` cutout

This is a follow-up to pull 148323. It mints
`-fsanitize-ignore-for-ubsan-feature=...`, accepting a list of (UBSan)
sanitizers that should not cause
`__has_feature(undefined_behavior_sanitizer)` to evaluate true.

fmayer commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -3973,6 +3980,10 @@ bool CompilerInvocation::ParseLangArgs(LangOptions &Opts, ArgList &Args,
+    parseSanitizerKinds(
+        "-fsanitize-ignore-for-ubsan-feature=",
+        Args.getAllArgValues(OPT_fsanitize_ignore_for_ubsan_feature_EQ), Diags,
+        Opts.UBSanFeatureSuppressedSanitize);
```
I would use consistent terminology regarding `ignore` or `suppress` for the flag and the opt

j39m replied:
Thanks --- fixed in a follow-up.

fmayer commented on clang/lib/Driver/SanitizerArgs.cpp:
```diff
@@ -736,6 +743,22 @@ SanitizerArgs::SanitizerArgs(const ToolChain &TC,
+  // in `SanitizerKind::Undefined`.
+  const SanitizerMask not_ubsan_mask =
+      IgnoreForUbsanFeature &
+      ~(SanitizerKind::Undefined | SanitizerKind::ShiftGroup);
```
hmmm. not sure this gymnastics are worth if for the warning

j39m replied:
Understood --- removed all this in the follow-up.

Commit:
fixup! Add fine-grained `__has_feature()` cutout

*   Uniformly use "ignore" rather than "suppress" in describing how we
    _ignore_ `-fsanitize=...` when evaluating `__has_feature()`.

*   Remove diagnostic emission when a sanitizer not belonging to UBSan
    is passed to `-fsanitize-ignore...`

Commit:
fixup! Add fine-grained `__has_feature()` cutout

Speculative fix for crash in
`Clang.Lexer/has_feature_undefined_behavior_sanitizer.cpp`. I can't
reproduce in my tree (am I building the right check target?), so this is
a guess.

Commit:
fixup! Add fine-grained `__has_feature()` cutout

Correct a typo: `RUN`, not `BUN`

Commit:
fixup! Add fine-grained `__has_feature()` cutout

Remove unused diagnostic.

j39m commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -277,6 +277,7 @@ def err_drv_malformed_sanitizer_metadata_ignorelist : Error<
   "malformed sanitizer metadata ignorelist: '%0'">;
 def err_drv_unsupported_static_sanitizer_darwin : Error<
   "static %0 runtime is not supported on darwin">;
+def err_drv_not_a_ubsan_sanitizer : Error<"not a UBSan sanitizer: '%0'">;
```
Thanks - fixed.

Commit:
fixup! Add fine-grained `__has_feature()` cutout

Reword comment for `LangOptions::UBSanFeatureIgnoredSanitize`.

vitalybuka reviewed (approved):
I suspect this new functionality is unnecessary micromanagement.
But I don't expect issues from having that.
So LGTM
